docs: add documentation about depinject and app wiring (#13683)
cosmos/cosmos-sdk: docs/docs/building-modules/14-depinject.md

---
sidebar_position: 1
---

# Dependency Injection

:::note

### Pre-requisite Readings

* [Cosmos SDK Dependency Injection Framework](../tooling/02-depinject.md)

:::

[`depinject`](../tooling/02-depinject.md) is used to wire any module in `app.go`.
All core modules are already configured to support dependency injection.

> Modules, to work with `depinject`, must defines its configuration and its requirements so that `depinject` can provide the right dependencies.

Reviewer: ```suggestion
To work with `depinject` a module must define its configuration and requirements so that `depinject` can provide the right dependencies.
```

In brief, as a module developer, the following steps are required:

1. Define the module configuration using Protobuf
2. Define the module dependencies in `x/{moduleName}/module.go`

A chain developer can then use the module by following these two steps:

> 1. Configure the module in `app_config.go`

Reviewer 2: ```suggestion
1. Configure the module in `app_config.go` or `app.yaml`
```

2. Inject the module in `app.go`

## Module Configuration

The module available configuration is defined in a Protobuf file, located at `{moduleName}/module/v1/module.proto`.

```protobuf reference
https://github.com/cosmos/cosmos-sdk/blob/main/proto/cosmos/group/module/v1/module.proto
```

* `go_import` must point to the Go package of the custom module.

> * Message fields defines the module configuration.

Reviewer: ```suggestion
* Message fields define the module configuration.
```

  That configuration can be set in the `app_config.go` / `app.yaml` file for a chain developer to configure the module.  

>   Taking `group` as example, a chain developer is able to decide, thanks to `uint64 max_metadata_len`, what is the maximum metatada length allowed for a group porposal.

Reviewer: ```suggestion
  Taking `group` as example, a chain developer is able to decide, thanks to `uint64 max_metadata_len`, what the maximum metatada length allowed for a group porposal is.
```

  ```go reference
    https://github.com/cosmos/cosmos-sdk/blob/0d8787c/simapp/app_config.go#L202-L206
  ```

> That message is generated using [`pulsar`](https://github.com/cosmos/cosmos-sdk/blob/main/scripts/protocgen-pulsar.sh) (by running `make proto-gen`).

Reviewer 2: Pulsar is totally optional. They can use the official `protoc-gen-go` if they like: https://developers.google.com/protocol-buffers/docs/reference/go-generated

Author: Clarified!

   * `depinject` provides the right dependencies to the module.
   * `depinject` also checks that all dependencies are provided.

  :::tip
  For making a dependency optional, add the `optional:"true"` struct tag.  
  :::

  ```go reference
  https://github.com/cosmos/cosmos-sdk/blob/main/x/group/module/module.go#L198-L208
  ```

4. Define the module outputs with a public struct that inherits `depinject.Out`:
   The module outputs are the dependencies that the module provides to other modules. It is usually the module itself and its keeper.

  ```go reference
  https://github.com/cosmos/cosmos-sdk/blob/main/x/group/module/module.go#L210-L215
  ```

5. Create a function named `ProvideModule` (as called in 1.) and use the inputs for instantitating the module outputs.

  ```go reference
  https://github.com/cosmos/cosmos-sdk/blob/main/x/group/module/module.go#L217-L227
  ```

Following is the complete app wiring configuration for `group`:

```go reference
https://github.com/cosmos/cosmos-sdk/blob/0d8787c/x/group/module/module.go#L180-L227
```

The module is now ready to be used with `depinject` by a chain developer.

## App Wiring

> The App Wiring is done in `app_config.go` / `app.yaml` and `app.go` and is explained in details in the [overview of `app.go`](../building-apps/00-app-go.md).

Reviewer: ```suggestion
The App Wiring is done in `app_config.go` / `app.yaml` and `app.go` and is explained in detail in the [overview of `app.go`](../building-apps/00-app-go.md).
```
